src/parser/SparqlParser.cpp

```diff
@@ -555,11 +555,43 @@ void SparqlParser::parseSolutionModifiers(ParsedQuery* query) {
                lexer_.peek("offset")) {
       query->_limitOffset =
           parseWithAntlr(sparqlParserHelpers::parseLimitOffsetClause, *query);
-    } else if (lexer_.accept(SparqlToken::Type::GROUP_BY)) {
-      lexer_.expect(SparqlToken::Type::VARIABLE);
-      query->_groupByVariables.emplace_back(lexer_.current().raw);
-      while (lexer_.accept(SparqlToken::Type::VARIABLE)) {
-        query->_groupByVariables.emplace_back(lexer_.current().raw);
+    } else if (lexer_.peek(SparqlToken::Type::GROUP_BY)) {
+      auto group_keys =
+          parseWithAntlr(sparqlParserHelpers::parseGroupClause, *query);
+
+      auto processVariableGroupKey = [&query](VariableGroupKey groupKey) {
```

Comment:
This whole following block basically looks the same as what we do for the `ORDER BY` keys, the common implementation should be factored out.